File: ios/Kwift/com/lightningkite/kwift/observables/Closeable.actual.swift

//--- Closeable
public protocol Closeable {
    func close()
}

//--- Close.{
public class Close: Closeable {
    
    let closer: ()->Void
    
    //--- Close.Primary Constructor
    public init(_ closer: @escaping () -> Void) {
        self.closer = closer
    }
    
    //--- Close.disposed (overwritten on flow generation)
    public var disposed: Bool {
        get {
            TODO()
        }
        set(value) {
            TODO()
        }
    }
    
    //--- Close.isDisposed() (overwritten on flow generation)
    public func isDisposed() -> Bool {
        TODO()
    }
    
    //--- Close.close()
    public func close() -> Void {
        closer()
    }
    
    //--- Close.}
}

File: android/src/main/java/com/lightningkite/khrysalis/observables/Closeable.actual.kt
package com.lightningkite.khrysalis.observables

import com.lightningkite.khrysalis.Escaping
import io.reactivex.disposables.Disposable

interface Closeable: Disposable {
    override fun isDisposed(): Boolean {
        return false
    }

    override fun dispose() {
        close()
    }

    fun close()
}

class Close(val closer: @Escaping() () -> Unit): Closeable {
    var disposed: Boolean = false
    override fun isDisposed(): Boolean {
        return disposed
    }

    override fun close() {
        disposed = true
        closer()
    }
}
